File: year2019/go/pkg/coders/intCoderDay11.go
package coders

import (
	"fmt"
	"math"
)

type Coordinate struct {
	x int
	y int
}

func NewCoordinate(x int, y int) Coordinate {
	return Coordinate{x, y}
}

type PaintHullIntCoder struct {
	IntCoder
}

func (ph *PaintHullIntCoder) OutputHullPaint(visited map[Coordinate]int, writer func(output []string)) {
	xMin, yMin := math.MaxInt, math.MaxInt
	xMax, yMax := math.MinInt, math.MinInt
	for coord := range visited {
		if coord.x > xMax {
			xMax = coord.x
		}
		if coord.y > yMax {
			yMax = coord.y
		}
		if coord.x < xMin {
			xMin = coord.x
		}
		if coord.y < yMin {
			yMin = coord.y
		}
	}
	var xLength = math.Abs(float64(xMin)) + math.Abs(float64(xMax)) + 1
	var yLength = math.Abs(float64(yMin)) + math.Abs(float64(yMax)) + 1
	rows := make([]string, int(yLength))
	var output string
	for i := 0; i < int(xLength); i++ {
		for j := 0; j < int(yLength); j++ {
			output = "."
			if visited[Coordinate{i - int(math.Abs(math.Abs(float64(xMin)))), j - int(math.Abs(math.Abs(float64(yMin))))}] == 1 {
				output = "#"
			}
			rows[j] = output
		}
		writer(rows)
		// fmt.Println(rows)
	}
}

func (ph *PaintHullIntCoder) PaintHull(codes []int) map[Coordinate]int {
	visited := map[Coordinate]int{}
	ph.PaintHullWithInput(codes, visited)
	return visited
}

func (ph *PaintHullIntCoder) findActionFromOutput(output int) int {
	switch output {
	case 0:
		return 0
	case 1:
		return 1
	}
	panic(fmt.Sprintf("Output not recognized: %d", output))
}

func (ph *PaintHullIntCoder) PaintHullWithInput(codesInput []int, visited map[Coordinate]int) {
	defer func() {
		ph.idx = 0
		ph.relativeBase = 0
	}()
	codes := make(map[int]int, len(codesInput))
	for i, v := range codesInput {
		codes[i] = v
	}
	currentPosition := Coordinate{0, 0}
	direction := Coordinate{0, 1}
	outputCallCount := 0
optLoop:
	for {
		execution := codes[ph.idx]
		switch optCode := execution % 100; optCode {
		case 1:
			codes[ph.getIdxFromMode(codes, execution, 3)] =
				codes[ph.getIdxFromMode(codes, execution, 2)] + codes[ph.getIdxFromMode(codes, execution, 1)]
			ph.idx += 4
		case 2:
			bar := codes[ph.getIdxFromMode(codes, execution, 2)] * codes[ph.getIdxFromMode(codes, execution, 1)]
			newIdx := ph.getIdxFromMode(codes, execution, 3)
			codes[newIdx] = bar

			ph.idx += 4
		case 3:
			codes[ph.getIdxFromMode(codes, execution, 1)] = visited[currentPosition]
			ph.idx += 2
		case 4:
			outputCallCount++
			output := codes[ph.getIdxFromMode(codes, execution, 1)]
			ph.idx += 2
			action := ph.findActionFromOutput(output)
			if outputCallCount%2 == 1 {
				visited[currentPosition] = action
				continue
			}
			switch action {
			case 0:
				direction = Coordinate{-direction.y, direction.x}
			case 1:
				direction = Coordinate{direction.y, -direction.x}
			}
			currentPosition = Coordinate{currentPosition.x + direction.x, currentPosition.y + direction.y}
		case 5:
			if codes[ph.getIdxFromMode(codes, execution, 1)] != 0 {
				ph.idx = codes[ph.getIdxFromMode(codes, execution, 2)]
				break
			}
			ph.idx += 3
		case 6:
			if codes[ph.getIdxFromMode(codes, execution, 1)] == 0 {
				ph.idx = codes[ph.getIdxFromMode(codes, execution, 2)]
				break
			}
			ph.idx += 3
		case 7:
			var toAssign int
			if codes[ph.getIdxFromMode(codes, execution, 1)] < codes[ph.getIdxFromMode(codes, execution, 2)] {
				toAssign = 1
			}
			codes[ph.getIdxFromMode(codes, execution, 3)] = toAssign
			ph.idx += 4
		case 8:
			var toAssign int
			if codes[ph.getIdxFromMode(codes, execution, 1)] == codes[ph.getIdxFromMode(codes, execution, 2)] {
				toAssign = 1
			}
			codes[ph.getIdxFromMode(codes, execution, 3)] = toAssign
			ph.idx += 4
		case 9:
			ph.relativeBase += codes[ph.getIdxFromMode(codes, execution, 1)]
			ph.idx += 2
		case 99:
			break optLoop
		default:
			panic(fmt.Sprintf("OptCode not known: %d", optCode))
		}
	}
}

File: year2019/cs/Common/IntCoderDay11.cs
namespace Common
{
    /// <summary>
    /// Used for Day 11.
    /// </summary>
    public partial class IntCoder
    {
        public void OutputHullPaint(
            DefaultDict<(int X, int Y), int> visited,
            Action<string> output)
        {
            (int xMin, int yMin, int xMax, int yMax) = (int.MaxValue, int.MaxValue, int.MinValue, int.MinValue);

            foreach (var (key, color) in visited)
            {
                if (key.X > xMax)
                {
                    xMax = key.X;
                }
                if (key.X < xMin)
                {
                    xMin = key.X;
                }
                if (key.Y > yMax)
                {
                    yMax = key.Y;
                }
                if (key.Y < yMin)
                {
                    yMin = key.Y;
                }
            }
            var xLength = Math.Abs(xMin) + Math.Abs(xMax) + 1;
            var yLength = Math.Abs(yMin) + Math.Abs(yMax) + 1;
            Span<char> rows = stackalloc char[yLength];
            for (var i = 0; i < xLength; i++)
            {
                for (var j = 0; j < yLength; j++)
                {
                    rows[j] = visited[(i - Math.Abs(xMin), j - Math.Abs(yMin))] == 1 ? '#' : '.';
                }
                output(rows.ToString());
            }
        }

        // Don't need visisted count - refactod
        public void PaintHullWithInput(IList<long> codesInput, DefaultDict<(int, int), int> visited)
        {
            (int X, int Y) currentPosition = (0, 0);
            (int X, int Y) direction = (0, 1);
            var outputCallCount = 0;
            var codes = new DefaultDict<long, long>();
            for (int i = 0; i < codesInput.Count; i++)
            {
                codes.Add(i, codesInput[i]);
            }
            try
            {
                do
                {
                    var execution = codes[idx];
                    switch (execution % 100)
                    {
                        case 1:
                            codes[GetIdxFromMode(codes, execution, 3)] =
                                codes[GetIdxFromMode(codes, execution, 2)] + codes[GetIdxFromMode(codes, execution, 1)];
                            idx += 4;
                            break;
                        case 2:
                            codes[GetIdxFromMode(codes, execution, 3)] =
                                codes[GetIdxFromMode(codes, execution, 2)] * codes[GetIdxFromMode(codes, execution, 1)];
                            idx += 4;
                            break;
                        case 3:
                            codes[GetIdxFromMode(codes, execution, 1)] = visited[currentPosition];
                            idx += 2;
                            break;
                        case 4:
                            outputCallCount++;
                            var output = codes[GetIdxFromMode(codes, execution, 1)];
                            idx += 2;
                            var action = output switch
                            {
                                0 => 0,
                                1 => 1,
                                _ => throw new Exception($"Output not correct: {output}")
                            };
                            if (outputCallCount % 2 == 1)
                            {
                                visited[currentPosition] = action;
                                continue;
                            }
                            switch (action)
                            {
                                case 0:
                                    direction = (-direction.Y, direction.X);
                                    break;
                                case 1:
                                    direction = (direction.Y, -direction.X);
                                    break;
                            }
                            currentPosition = (currentPosition.X + direction.X, currentPosition.Y + direction.Y);
                            break;
                        case 5:
                            if (codes[GetIdxFromMode(codes, execution, 1)] != 0)
                            {
                                idx = codes[GetIdxFromMode(codes, execution, 2)];
                                break;
                            }
                            idx += 3;
                            break;
                        case 6:
                            if (codes[GetIdxFromMode(codes, execution, 1)] == 0)
                            {
                                idx = codes[GetIdxFromMode(codes, execution, 2)];
                                break;
                            }
                            idx += 3;
                            break;
                        case 7:
                            codes[GetIdxFromMode(codes, execution, 3)] =
                                codes[GetIdxFromMode(codes, execution, 1)] < codes[GetIdxFromMode(codes, execution, 2)] ?
                                1 : 0;
                            idx += 4;
                            break;
                        case 8:
                            codes[GetIdxFromMode(codes, execution, 3)] =
                                codes[GetIdxFromMode(codes, execution, 1)] == codes[GetIdxFromMode(codes, execution, 2)] ?
                                1 : 0;
                            idx += 4;
                            break;
                        case 9:
                            relativeBase += codes[GetIdxFromMode(codes, execution, 1)];
                            idx += 2;
                            break;
                        case 99:
                            idx = codes.Count;
                            break;
                        default:
                            throw new Exception($"OptCode not known {execution}");
                    }
                } while (idx < codes.Count);
            }
            finally
            {
                idx = 0;
                relativeBase = 0;
            }
        }

        public DefaultDict<(int, int), int> PaintHull(IList<long> codesInput)
        {
            var visited = new DefaultDict<(int, int), int>();
            PaintHullWithInput(codesInput, visited);
            return visited;
        }
    }
}
